homeassistant/components/velbus.py

```diff
@@ -26,18 +28,26 @@
 }, extra=vol.ALLOW_EXTRA)
 
 
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Set up the Velbus platform."""
     import velbus
     port = config[DOMAIN].get(CONF_PORT)
-    connection = velbus.VelbusUSBConnection(port)
-    controller = velbus.Controller(connection)
+    controller = velbus.Controller(port)
     hass.data[DOMAIN] = controller
 
     def stop_velbus(event):
         """Disconnect from serial port."""
         _LOGGER.debug("Shutting down ")
-        connection.stop()
+        controller.stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_velbus)
+
+    def callback():
+        """Callback when scan is completed."""
+        hass.async_add_job(async_load_platform(hass, 'switch', DOMAIN))
+        hass.async_add_job(async_load_platform(hass, 'binary_sensor', DOMAIN))
+
+    controller.scan(callback)
```

Comment:
This call contains a `time.sleep` (3 seconds) call. Sleeping is not allowed inside the event loop. Doing it in a worker thread is not optimal either, since if that would be done frequently, the thread pool would be exhausted. Is there another option than sleeping?

Reply:
@MartinHjelmare Thank you for your comments. I'm working on incorporating these in the PR.

On the timer: time.sleep is not executed in the event loop but in a thread (see __init__.py in connections directory: the message and callback is put in a write queue that is read by the write thread.

I don't see any alternative to a sleep command given that there is no way to know which addresses on the bus are being used, hence I iterate over all addresses, wait for a couple of seconds and see which ones respond and which ones don't. Happy to get feedback if you would see an alternative solution.

Reply:
Sleeping in a separate thread is fine. :+1: